File: algoritms/binary-search/binary-search-last-occurrence.c
#include <math.h>

int binarySearchForLastOccurrence(int* nums, int numsSize, int target) {
    int low = 0;
    int high = numsSize - 1;

    while (low <= high) {
        double midIndex = low + (high - low) / 2.0;
        int mid = ceil(midIndex);

        if (nums[mid] < target) {
            low = mid + 1;
            continue;
        }

        if (nums[mid] > target) {
            high = mid - 1;
            continue;
        }

        if (high != mid) {
            low = mid;
            continue;
        }

        return mid;
    }

    return -1;
}

File: algoritms/binary-search/binary-search-last-occurrence.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number} index or -1 if nothing found
 */
var binarySearchLastOccurrence = function (nums, target) {
    let low = 0;
    let high = nums.length - 1;

    while (low <= high) {
        const mid = Math.ceil(low + (high - low) / 2)
        const value = nums[mid]
       
        if (value < target) {
            low = mid + 1;
            continue;
        }

        if (value > target) {
            high = mid - 1;
            continue;
        }

        if (high != mid) {
            low = mid;
            continue;
        }

        return mid
    }

    return -1
}
